data/is_ocaml_yet/web.md

```diff
@@ -0,0 +1,258 @@
+---
+id: web
+question: Is OCaml web yet?
+answer: Not quite yet, but we're getting there!
+categories:
+  - name: Authentication
+    status: 🆘
+    description: |
+      n/a
+    packages:
+    - FPauth
+    - u2f
+    - webauthn
+    - jwt
+    - cookie
+    - ssl
+    - twostep
+    - oidc
+  - name: Binary protocols
+    status: 🟡
+    description: |
+      There are implementations for a lot of popular protocols, some of them are solid, but some of them are still incomplete and need more testing. Some useful protocols for which there is no package include FlatBuffers and AMQP.
+    packages:
+    - asn1-combinators
+    - avro
+    - capnp
+    - grpc
+    - protobuf
+    - rpc
+    - jsonrpc
+    - thrift
+    - bson
+    - cbor
+    - mqtt
+  - name: Browser
+    status: 🟢
+    description: |
+      OCaml has excellent support for running in the browser!
+      The [Js_of_ocaml](/p/js_of_ocaml/latest) compiler, which translates OCaml into JavaScript, is a well-established tool dating back to 2010. It integrates seamlessly with the existing OCaml ecosystem and powers numerous production applications.
+
+      In addition, [Melange](/p/melange/latest) is a more recent entry to the OCaml-JavaScript compiler space. Originating as a fork of [ReScript](https://rescript-lang.org/), Melange may not have the extensive track record of Js_of_ocaml, but it builds on a very strong foundation and already integrates well with Dune, the OCaml build system.
+    packages:
+    - js_of_ocaml
+    - melange
+    - brr
+    - gen_js_api
+  - name: Compression
+    status: 🟡
+    description: |
+      n/a
+    packages:
+    - zlib
+    - decompress
+    - camlzip
+    - lz4
+    - ezgzip
+    - hpack
+    - brotli
+    - zstd
+    - snappy
+    - lzo
+    - zstandard
+  - name: Crypto
+    status: 🟡
+    description: |
+      n/a
+    packages:
+    - cryptokit
+    - mirage-crypto
+    - nocrypto
+    - pkcs11
+    - tls
+    - hacl_x25519
+    - x509
+    - fiat-p256
+    - sha
+    - blake2
+    - blake3
+    - bls12-381
+    - pbkdf
+    - dirsp-proscript
+    - digestif
+  - name: Database
+    status: 🟠
+    description: |
+      n/a
+    packages:
+    - caqti
+    - pgocaml
+    - sqlite3
+    - lmdb
+    - irmin
+    - mysql
+    - postgresql
+    - aws-rds
+    - sequoia
+    - sqlgg
+    - petrol
+    - mariadb
+  - name: Email
+    status: 🟠
+    description: |
+      n/a
+    packages:
+    - tidy_email
+    - emile
+    - smtp
+    - letters
+    - mrmime
+    - sendmail
+    - facteur
+    - received
+    - email_message
+    - colombe
+  - name: En- & Decoding
+    status: 🟠
+    description: |
+      n/a
+    packages:
+    - tar
+    - imagelib
+    - base64
+    - multipart-form-data
+    - multipart_form
+    - biniou
+  - name: Web Frameworks
+    status: 🟠
+    description: |
+      n/a
+    packages:
+    - dream
+    - eliom
+    - opium
+    - sihl
+  - name: HTTP Clients
```

Comment:
See also https://discuss.ocaml.org/t/simple-modern-http-client-library/11239/14?u=edwin
The short version is that unfortunately if you want something that works with both HTTP/1.x and HTTP/2 and is easy to use, then your best bet is using a binding to a C library (ocurl).
If you're willing to use only HTTP/1.x or only HTTP/2.x or are fine with some of the other limitations of the pure OCaml implementations then there are many choices available (problem there being the many choices, you can spend a large amount of time just testing each until you find a combination that works for you, and inevitably there'll be bugs when all you wanted to do is GET a url or POST a URL).
Hopefully this can be improved, I think all the pieces are there, they are just not the most easy to use or come with a very large dependency cone.

The situation is much better when writing an HTTP *server* though.

Reply:
FWIW, your table brought me to revising reverse dependencies in the mirage-crypto / ocaml-tls / dns-client environment, and I suspect the "OCaml Dependencies" and "System Dependencies" may need an update (also, I feel that a brief number is only about quantity about packages - but you can always bundle everything and get to a single dependency -- so maybe it's not the healthiest metrics (since people then strive for few large dependencies)). In any case, thanks again for the table which motivated me to disentangle some code.